diff -urN gcc-3.3.6/configure.in gcc-3.3.6-new/configure.in
--- gcc-3.3.6/configure.in	2004-01-02 17:09:48.000000000 +0300
+++ gcc-3.3.6-new/configure.in	2009-02-10 18:08:43.000000000 +0300
@@ -638,6 +638,9 @@
   *-*-lynxos*)
     noconfigdirs="$noconfigdirs target-newlib target-libgloss ${libgcj}"
     ;; 
+  msp430-*-*)
+    noconfigdirs="$noconfigdirs ${libstdcxx_version}"
+    ;;
   *-*-*)
     noconfigdirs="$noconfigdirs ${libgcj}"
     ;;
diff -urN gcc-3.3.6/gcc/collect2.c gcc-3.3.6-new/gcc/collect2.c
--- gcc-3.3.6/gcc/collect2.c	2003-12-08 22:02:39.000000000 +0300
+++ gcc-3.3.6-new/gcc/collect2.c	2009-02-10 18:09:09.000000000 +0300
@@ -1572,7 +1572,7 @@
   if (redir)
     {
       /* Open response file.  */
-      redir_handle = open (redir, O_WRONLY | O_TRUNC | O_CREAT);
+      redir_handle = open (redir, O_WRONLY | O_TRUNC | O_CREAT, 0755);
 
       /* Duplicate the stdout and stderr file handles
 	 so they can be restored later.  */
